### PR TITLE
Add CMK support for core resources

### PR DESCRIPTION
**Pull request by yuvalyaron**
# Resolves https://github.com/microsoft/AzureTRE/issues/4142, https://github.com/microsoft/AzureTRE/issues/4143
Part of #4002 

## What is being addressed

Added CMK support for Storage accounts and VM storage disks for core TRE resources.
Added the option to use an external Key Vault to store the CMKs

## How is this addressed

users set `enable_cmk_encryption: true` in the `config.yaml`,
and either `encryption_kv_name` or `external_key_store_id` for the key vault.

### CODE REVIEW

File: core/terraform/airlock/airlock_processor.tf [hunk 1]
```diff
@@ -27,9 +27,31 @@ resource "azurerm_storage_account" "sa_airlock_processor_func_app" {
...
 }
 
+resource "azurerm_storage_account_customer_managed_key" "sa_airlock_processor_func_app_encryption" {
+  count                     = var.enable_cmk_encryption ? 1 : 0
+  storage_account_id        = azurerm_storage_account.sa_airlock_processor_func_app.id
+  key_vault_id              = var.key_store_id
+  key_name                  = var.kv_encryption_key_name
+  user_assigned_identity_id = var.encryption_identity_id
+
+  lifecycle {
```
Comment (guybartal) [inline core/terraform/airlock/airlock_processor.tf:48]: is that a requirement not to recreate the CMK when KV changes?

File: core/terraform/resource_processor/vmss_porter/main.tf [hunk 2]
```diff
@@ -50,6 +50,25 @@ resource "azurerm_user_assigned_identity" "vmss_msi" {
   lifecycle { ignore_changes = [tags] }
 }
 
+resource "azurerm_disk_encryption_set" "vmss_disk_encryption" {
+  count                     = var.enable_cmk_encryption ? 1 : 0
+  name                      = "vmss-disk-encryption-rp-porter-${var.tre_id}"
```
Comment (guybartal) [inline core/terraform/resource_processor/vmss_porter/main.tf:55]: make sure name is not too long

File: devops/terraform/cmk_encryption.tf [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+# Key used to encrypt management resources
+resource "azurerm_key_vault_key" "tre_mgmt_encryption" {
+  count = var.enable_cmk_encryption ? 1 : 0
+
+  name         = var.kv_mgmt_encryption_key_name
+  key_vault_id = local.key_store_id
+  key_type     = "RSA"
+  key_size     = 2048
+
+  key_opts = [
```
Comment (guybartal) [inline devops/terraform/cmk_encryption.tf:27]: nit: if the key is not used only for encryption / decryption, consider naming the module not cmk_encyption but more generic name

**Review by guybartal — APPROVED** [review]
LGTM with few comments

File: core/terraform/airlock/airlock_processor.tf (hunk 1 above)
Comment (yuvalyaron) [reply core/terraform/airlock/airlock_processor.tf]: thanks, removed

File: devops/terraform/cmk_encryption.tf (hunk 3 above)
Comment (yuvalyaron) [reply devops/terraform/cmk_encryption.tf:27]: Thanks, removed the unused sign and verify key options.
This key is solely for encryption.

File: config.sample.yaml [hunk 4]
```diff
@@ -8,6 +8,10 @@ management:
   mgmt_storage_account_name: __CHANGE_ME__
   terraform_state_container_name: tfstate
   acr_name: __CHANGE_ME__
+  # ID of external Key Vault to store CMKs in (only required if enable_cmk_encryption is true)
```
Comment (LizaShak) [inline config.sample.yaml:11]: I think worth adding to documentation as well : https://microsoft.github.io/AzureTRE/latest/tre-admins/environment-variables/

**Review by LizaShak — APPROVED** [review]
Looks good, just a small comment to document all the new config

File: config.sample.yaml (hunk 4 above)
Comment (yuvalyaron) [reply config.sample.yaml:11]: thanks! added

File: docs/tre-admins/environment-variables.md [hunk 5]
```diff
@@ -42,7 +44,7 @@
 | `FIREWALL_SKU` | Optional. The SKU of the Azure Firewall instance. Default value is `Standard`. Allowed values [`Basic`, `Standard`, `Premium`]. See [Azure Firewall SKU feature comparison](https://learn.microsoft.com/en-us/azure/firewall/choose-firewall-sku). |
 | `APP_GATEWAY_SKU` | Optional. The SKU of the Application Gateway. Default value is `Standard_v2`. Allowed values [`Standard_v2`, `WAF_v2`] |
 | `CUSTOM_DOMAIN` | Optional. Custom domain name to access the Azure TRE portal. See [Custom domain name](custom-domain.md). |
-
+| `ENABLE_CMK_ENCRYPTION` | If set to `true`, customer-managed key encryption will be enabled for all supported resources. |
```
Comment (marrobi) [inline docs/tre-admins/environment-variables.md:47]: @yuvalyaron is there additional configuration required - such as KeyVault etc? Would be good to have something in the docs detailing how we actually make this "work". Thanks.

Comment (marrobi) [reply docs/tre-admins/environment-variables.md:47]: These are also env vars?

```hcl
variable "enable_cmk_encryption" {
  type        = bool
  description = "A boolean indicating if customer managed keys will be used for encryption of supporting resources"
  default     = false
}

variable "external_key_store_id" {
  type        = string
  description = "ID of external Key Vault to store CMKs in (only required if enable_cmk_encryption is true)"
  default     = null
}

variable "encryption_kv_name" {
  type        = string
  description = "Name of Key Vault for encryption keys, required only if external_key_store_id is not set (only used if enable_cmk_encryption is true)"
  default     = null
}

variable "kv_encryption_key_name" {
  type        = string
  description = "Name of Key Vault Encryption Key (only used if enable_cmk_encryption is true)"
  default     = "tre-encryption"
}
```

Comment (yuvalyaron) [reply docs/tre-admins/environment-variables.md:47]: - I'll add a section in the docs.
- The env vars you mentioned are documented here

Comment (yuvalyaron) [reply docs/tre-admins/environment-variables.md:47]: added a section in the docs

File: config.sample.yaml [hunk 6]
```diff
@@ -80,6 +84,9 @@ developer_settings:
 # the base workspace.
 #  enable_local_debugging: true
 
+# This setting enables customer-managed key encryption for all supported resources
+#  enable_cmk_encryption: true
```
Comment (tim-allen-ck) [inline config.sample.yaml:88]: Is the default value false?

**Comment by tim-allen-ck** [comment]
Has support been adding for the CMK configuration through Env Vars in GitHub?

File: config.sample.yaml (hunk 6 above)
Comment (yuvalyaron) [reply config.sample.yaml:88]: yes

**Comment by yuvalyaron** [comment]
> Has support been adding for the CMK configuration through Env Vars in GitHub?

https://github.com/microsoft/AzureTRE/issues/4148
@tim-allen-ck

File: config.sample.yaml (hunk 6 above)
Comment (tim-allen-ck) [reply config.sample.yaml:88]: could that be said explicit somewhere?

**Review by tim-allen-ck — APPROVED** [review]
LGTM

**Pull request merged**